File: 백준/Silver/1389. 케빈 베이컨의 6단계 법칙/케빈 베이컨의 6단계 법칙.py
# 케빈 베이컨의 6단계 법칙
import sys
input = sys.stdin.readline
from collections import deque

def bfs(n):
    que = deque([(n, 0)])
    result = 0
    while que:
        cur, cnt = que.popleft()

        if visited[cur]:
            continue
        else:
            visited[cur] = 1
            result += cnt

        for i in graph[cur]:
            if not visited[i]:
                que.append((i, cnt+1))
    return result
        

n, m = map(int,input().split())
graph = [[] for _ in range(n+1)]
minn = float('inf')
result = 0

for i in range(m):
    s, t = map(int,input().split())
    graph[s].append(t)
    graph[t].append(s)

for i in range(1, n+1):
    visited = [0] * (n+1)
    cur = bfs(i)
    if minn > cur:
        result = i
        minn = cur

print(result)

File: 백준/Silver/1389. 케빈 베이컨의 6단계 법칙/케빈 베이컨의 6단계 법칙.java
import java.io.*;
import java.util.*;

public class Main {
    static boolean[] visited;
    static ArrayList<Integer>[] graph;
    static int bfs(int n) {
        ArrayDeque<int[]> que = new ArrayDeque<>();
        que.addLast(new int[]{n, 0});
        int result = 0;

        while (!que.isEmpty()) {
            int[] cur = que.removeFirst();
            int node = cur[0];
            int cnt = cur[1];

            if (visited[node]) {
                continue;
            } else {
                visited[node] = true;
                result += cnt;
            }

            for (int i : graph[node]) {
                if (!visited[i]) {
                    que.addLast(new int[]{i, cnt+1});
                }
            }
        }
        return result;
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int result = 0;
        int min = Integer.MAX_VALUE;

        graph = new ArrayList[n+1];
        for (int i = 0; i < n+1; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int s = Integer.parseInt(st.nextToken());
            int t = Integer.parseInt(st.nextToken());
            graph[s].add(t);
            graph[t].add(s);
        }

        for (int i = 1; i < n+1; i++) {
            visited = new boolean[n+1];
            int cur = bfs(i);

            if (min > cur) {
                result = i;
                min = cur;
            }
        }
        System.out.println(result);
    }
}
